File: Algorithms/Greedy/Reverse Shuffle Merge.php
<?php

/*
 * Complete the 'reverseShuffleMerge' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts STRING s as parameter.
 */

function reverseShuffleMerge($s) {
    # Write your code here
    $char_count = array();
    $used_chars = array();
    $remaining_chars = array();
    $result = array();

    for ($i = 0; $i < strlen($s); $i++) {
        $char = $s[$i];
        if (isset($char_count[$char])) {
            $char_count[$char]++;
        }
        else {
            $char_count[$char] = 1;
        }
        $remaining_chars[$char] = $char_count[$char];
    }

    function can_use($char, $char_count, $used_chars) {
        return (!isset($used_chars[$char]) || $used_chars[$char] < $char_count[$char] / 2);
    }

    function can_pop($char, $remaining_chars, $used_chars, $char_count) {
        $needed_chars = $char_count[$char] / 2 - (isset($used_chars[$char]) ? $used_chars[$char] : 0);
        return ($remaining_chars[$char] > $needed_chars);
    }

    for ($i = strlen($s) - 1; $i >= 0; $i--) {
        $char = $s[$i];

        if (can_use($char, $char_count, $used_chars)) {
            while (!empty($result) && end($result) > $char && can_pop(end($result), $remaining_chars, $used_chars, $char_count)) {
                $removed_char = array_pop($result);
                $used_chars[$removed_char]--;
            }

            $result[] = $char;
            if (isset($used_chars[$char])) {
                $used_chars[$char]++;
            }
            else {
                $used_chars[$char] = 1;
            }
        }

        $remaining_chars[$char]--;
    }

    return implode('', $result);$char_count = array();
    $used_chars = array();
    $remaining_chars = array();
    $result = array();

    for ($i = 0; $i < strlen($s); $i++) {
        $char = $s[$i];
        if (isset($char_count[$char])) {
            $char_count[$char]++;
        }
        else {
            $char_count[$char] = 1;
        }
        $remaining_chars[$char] = $char_count[$char];
    }

    function can_use($char, $char_count, $used_chars) {
        return (!isset($used_chars[$char]) || $used_chars[$char] < $char_count[$char] / 2);
    }

    function can_pop($char, $remaining_chars, $used_chars, $char_count) {
        $needed_chars = $char_count[$char] / 2 - (isset($used_chars[$char]) ? $used_chars[$char] : 0);
        return ($remaining_chars[$char] > $needed_chars);
    }

    for ($i = strlen($s) - 1; $i >= 0; $i--) {
        $char = $s[$i];

        if (can_use($char, $char_count, $used_chars)) {
            while (!empty($result) && end($result) > $char && can_pop(end($result), $remaining_chars, $used_chars, $char_count)) {
                $removed_char = array_pop($result);
                $used_chars[$removed_char]--;
            }

            $result[] = $char;
            if (isset($used_chars[$char])) {
                $used_chars[$char]++;
            }
            else {
                $used_chars[$char] = 1;
            }
        }

        $remaining_chars[$char]--;
    }

    return implode('', $result);
}

$fptr = fopen(getenv("OUTPUT_PATH"), "w");

$s = rtrim(fgets(STDIN), "\r\n");

$result = reverseShuffleMerge($s);

fwrite($fptr, $result. "\n");
fclose($fptr);

File: Algorithms/Greedy/Reverse Shuffle Merge.py
import os

#
# Complete the 'reverseShuffleMerge' function below.
#
# The function is expected to return a STRING.
# The function accepts STRING s as parameter.
#

def reverseShuffleMerge(s):
    # Write your code here
    char_count = {}
    for char in s:
        if char in char_count:
            char_count[char] += 1
        else:
            char_count[char] = 1

    used_chars = {}
    remaining_chars = dict(char_count)
    result = []

    def can_use(char):
        return used_chars.get(char, 0) < char_count[char] // 2

    def can_pop(char):
        needed_chars = char_count[char] // 2 - used_chars.get(char, 0)
        return remaining_chars[char] > needed_chars

    for char in reversed(s):
        if can_use(char):
            while result and result[-1] > char and can_pop(result[-1]):
                removed_char = result.pop()
                used_chars[removed_char] -= 1

            result.append(char)
            used_chars[char] = used_chars.get(char, 0) + 1

        remaining_chars[char] -= 1

    return "".join(result)


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    s = input()

    result = reverseShuffleMerge(s)

    fptr.write(result + '\n')
    fptr.close()
